### PR TITLE
SWIFT-244 TopologyDescription.hasReadableServer should take in a Read…

### PR DESCRIPTION
**Pull request by patrickfreed**
[SWIFT-244](https://jira.mongodb.org/browse/SWIFT-244)

This PR adds an optional read preference to `TopologyDescription.hasReadableServer`. 

In order to fully support the max staleness option of read preferences, we needed to basically perform server selection, which relies on two fields that libmongoc does not expose: `lastUpdateTime` from `mongoc_server_description_t` and `heartbeatFrequencyMS` from `mongoc_topology_description_t`. For the former, I filed a ticket (https://jira.mongodb.org/browse/CDRIVER-2896) to add the necessary getter. In the meantime, it can be estimated using the system clock in the Swift driver. The latter has a spec-defined default value and is configured via the connection string, so I just set it from there.

### CODE REVIEW

File: Sources/MongoSwift/SDAM.swift [hunk 1]
```diff
@@ -167,9 +162,35 @@ public struct ServerDescription {
...
+        self.roundTripTime = rtt
+        self.parseIsMaster(isMaster)
+        self.type = type
+        self.lastUpdateTime = updateTime // TODO: get lastUpdateTime from mongoc_server_description_t after CDRIVER-2896
+    }
+
     /// An internal initializer to create a `ServerDescription` from an OpaquePointer to a
     /// mongoc_server_description_t.
-    internal init(_ description: OpaquePointer) {
+    internal init(_ description: OpaquePointer, _ updateTime: Date?) {
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: Done and done

File: Sources/MongoSwift/SDAM.swift [hunk 2]
```diff
@@ -167,9 +162,35 @@ public struct ServerDescription {
...
+            if !serverTags.keys.contains(kvp.key) || !bsonEquals(serverTags[kvp.key]!, kvp.value) {
+                return false
+            }
+        }
+
+        return true
+    }
+
+    /// An internal initializer to create a `ServerDescription` for testing purposes.
+    internal init(_ connectionId: ConnectionId, _ rtt: Int64, _ isMaster: Document, _ type: ServerType,
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: 1. Done
2. You can only define convenience initializers in class extensions, so moving this initializer to the tests package won't work unfortunately

File: Sources/MongoSwift/SDAM.swift [hunk 3]
```diff
@@ -231,29 +255,110 @@ public struct TopologyDescription {
...
+        }
     }
 
     /// Returns `true` if the topology has a writable server available, and `false` otherwise.
     public func hasWritableServer() -> Bool {
         return [.single, .replicaSetWithPrimary].contains(self.type)
     }
 
+    /// An internal initializer to create a `TopologyDescription` for testing purposes
+    internal init(_ type: TopologyType, _ servers: [ServerDescription]) {
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: Labels added, see my comment above about moving the init to an extension.

File: Sources/MongoSwift/SDAM.swift [hunk 4]
```diff
@@ -231,29 +255,110 @@ public struct TopologyDescription {
...
+        }
+
+        return candidates.count > 0
+    }
+
     /// Returns `true` if the topology has a readable server available, and `false` otherwise.
-    public func hasReadableServer() -> Bool {
-        // (this function should take in an optional ReadPreference, but we have yet to implement that type.)
-        return [.single, .replicaSetWithPrimary, .sharded].contains(self.type)
+    public func hasReadableServer(_ readPref: ReadPreference = ReadPreference(ReadPreference.Mode.primary)) -> Bool {
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: Yep, shortened this.

File: Tests/MongoSwiftTests/SDAMMonitoringTests.swift [hunk 5]
```diff
@@ -118,4 +118,113 @@ final class SDAMTests: MongoSwiftTestCase {
...
+                    "hasReadableServer to return \(avail)")
+
+            if readPref.mode == .primary {
+                primaryCase = avail
+            }
+        }
+        expect(topology.hasReadableServer()).to(equal(primaryCase))
+    }
+
+    func testHasReadableServers() throws {
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Tests/MongoSwiftTests/SDAMMonitoringTests.swift]: Added

File: Sources/MongoSwift/SDAM.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (kmahar) [reply Sources/MongoSwift/SDAM.swift]: ugh! ok, that's fine. let's just add an extra line to the docstring here explaining why this test initializer lives here, because otherwise we will probably forget and try to move them again in the future :)

File: Sources/MongoSwift/SDAM.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (kmahar) [reply Sources/MongoSwift/SDAM.swift]: same reply as above, let's mention in docstring why it's defined here.

File: Sources/MongoSwift/SDAM.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (kmahar) [reply Sources/MongoSwift/SDAM.swift]: but hmm, wait. we used to have an initializer in [this extension of `Document`](https://github.com/mongodb/mongo-swift-driver/blob/aeb8e390640a65c1f1a7be8ad4e67b4cecabe76a/Sources/MongoSwift/BSON/Document.swift#L380).
perhaps it's something about defining it within a different package?

[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: Oh, you're right!
https://docs.swift.org/swift-book/LanguageGuide/Extensions.html#ID153 says explicitly for _classes_, not structs/value types, which `Document`, `TopologyDescription`, and `ServerDescription` all are.

But yes, I can't put the initializers in the test file due to it being in a separate package. I can move them around within MongoSwift though.

[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: So after some more digging, what we want actually _is_ possible in Swift 4, but the compiler will give a warning. In Swift 5, that warning will be an error (https://github.com/apple/swift-evolution/blob/master/proposals/0189-restrict-cross-module-struct-initializers.md). 

The reason all this confusion started is I read the aforementioned warning as an error and skimmed over the actual, un/kinda related one in the build output. 

What I'm not entirely clear on is why our tests can access functions marked `internal`, yet are considered outside of the module when implementing the initializers. Are our tests in some half-in-half-out state?

[the remark this answers is not in the document]
Comment (kmahar) [reply Sources/MongoSwift/SDAM.swift]: ah, I see. if it's going to be forbidden in Swift 5, then I don't think there is a point in doing it now either. 
and ha, we actually had a discussion about this in #swift last week. the reason we have that capability is in the import statement: note that our tests do `@testable import MongoSwift`, which elevates our access privileges.

from [docs](https://docs.swift.org/swift-book/LanguageGuide/AccessControl.html):
>  When you write an app with a unit test target, the code in your app needs to be made available to that module in order to be tested. By default, only entities marked as open or public are accessible to other modules. However, a unit test target can access any internal entity, if you mark the import declaration for a product module with the @testable attribute and compile that product module with testing enabled.

[the last section](https://github.com/apple/swift-evolution/blob/master/proposals/0189-restrict-cross-module-struct-initializers.md#add-an-exception-for-unit-tests) of the proposal you linked notes that they considered making an exception and allowing this when the module is imported via `@testable`, but that they ultimately chose not to since it can already be accomplished in exactly the way you've done it here. 

so yeah, it sounds like we should just leave them where they are, and briefly note why they're there in docstring.

[the remark this answers is not in the document]
Comment (kmahar) [reply Sources/MongoSwift/SDAM.swift]: (I would also explicitly mention SE-0189)

[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: Of course the one paragraph of that proposal I skipped was the most relevant...
Anyways, sounds good. Ill update the docstrings.

File: Sources/MongoSwift/SDAM.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: Done.

File: Sources/MongoSwift/SDAM.swift [hunk 6]
```diff
@@ -231,29 +254,111 @@ public struct TopologyDescription {
...
+            suitableModes = [ServerType.rsSecondary]
+        default:
+            suitableModes = [ServerType.rsSecondary, ServerType.rsPrimary]
+        }
+
+        var candidates: [ServerDescription] = servers.filter { suitableModes.contains($0.type) }
+
+        // If read preference specifies a max staleness, first filter out too stale servers
+        if let maxStalenessSeconds = readPref.maxStalenessSeconds {
+            candidates = candidates.filter { (candidate: ServerDescription) in
```
Comment (kmahar) [inline Sources/MongoSwift/SDAM.swift]: can probably omit type annotation for candidate here

File: Sources/MongoSwift/SDAM.swift [hunk 7]
```diff
@@ -231,29 +254,111 @@ public struct TopologyDescription {
...
+                let staleness = { () -> Double in
+                    let delta = { (serverDesc: ServerDescription) -> Double in
+                        (serverDesc.lastUpdateTime!.timeIntervalSinceReferenceDate -
```
Comment (kmahar) [inline Sources/MongoSwift/SDAM.swift]: we're doing a lot of force unwrapping of optionals throughout this block. generally, we only want to do it if we are sure that the values will be present, because otherwise Swift crashes. 
are we sure that they will be present? does the spec define behavior for cases where they are not?

Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: `lastUpdateTime` is not necessarily present, so I've added protections for when it isn't. 
`lastWriteDate` should always be present in the `ismaster` for secondaries and primaries, and all candidates are filtered to be primaries or secondaries. If you want me to make this check more explicit, let me know.

File: Sources/MongoSwift/SDAM.swift (hunk 6 above)
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: done

File: Sources/MongoSwift/SDAM.swift [hunk 8]
```diff
@@ -231,29 +254,111 @@ public struct TopologyDescription {
...
+        switch readPref.mode {
+        case ReadPreference.Mode.primary:
+            suitableModes = [ServerType.rsPrimary]
+        case ReadPreference.Mode.secondary:
+            suitableModes = [ServerType.rsSecondary]
+        default:
+            suitableModes = [ServerType.rsSecondary, ServerType.rsPrimary]
+        }
+
+        var candidates: [ServerDescription] = servers.filter { suitableModes.contains($0.type) }
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: Sounds good. Done and done

File: Sources/MongoSwift/SDAM.swift [hunk 9]
```diff
@@ -231,29 +254,111 @@ public struct TopologyDescription {
         }
     }
 
+    /// An internal function used to determine if there are suitable servers to read from in a replica set.
+    internal func hasSuitableReadServer(_ readPref: ReadPreference) -> Bool {
+        var suitableModes: [ServerType]
+
+        switch readPref.mode {
+        case ReadPreference.Mode.primary:
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: 👍

File: Sources/MongoSwift/SDAM.swift [hunk 10]
```diff
@@ -231,29 +254,111 @@ public struct TopologyDescription {
...
+            return false
+        case .single, .sharded:
+            return true
+        case .replicaSetNoPrimary:
+            if readPref.mode == ReadPreference.Mode.primary {
+                return false
+            }
+            return hasSuitableReadServer(readPref)
+        case .replicaSetWithPrimary:
+            if readPref.mode == ReadPreference.Mode.primary {
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: 👍

File: Sources/MongoSwift/SDAM.swift (hunk 7 above)
Comment (kmahar) [reply Sources/MongoSwift/SDAM.swift]: perfect - let's just add a comment somewhere around here explaining then that rationale.

Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: Done

File: Sources/MongoSwift/SDAM.swift [hunk 11]
```diff
@@ -231,29 +254,137 @@ public struct TopologyDescription {
...
+                return (serverLastUpdate.timeIntervalSinceReferenceDate -
+                        serverDesc.lastWriteDate!.timeIntervalSinceReferenceDate) * 1000.0
+            }
+
+            guard let clientToServer = delta(server), let clientToPrimary = delta(primary) else {
+                return nil
+            }
+
+            return clientToServer - clientToPrimary + Double(TopologyDescription.heartbeatFrequencyMS)
+        } else {
```
Comment (kmahar) [inline Sources/MongoSwift/SDAM.swift]: this is the replica set no primary case, right? (let's add comment)

File: Sources/MongoSwift/SDAM.swift [hunk 12]
```diff
@@ -231,29 +254,137 @@ public struct TopologyDescription {
...
+            let possibleSMax = self.servers.filter { $0.type == .rsSecondary }.max {
+                $0.lastWriteDate!.timeIntervalSinceReferenceDate < $1.lastWriteDate!.timeIntervalSinceReferenceDate
+            }
+
+            guard let sMax = possibleSMax else {
```
Comment (kmahar) [inline Sources/MongoSwift/SDAM.swift]: so this happens if there ended up being no secondaries, right? can we check for that before we call max? otherwise it's a little hard to parse what edge case this is checking for.

Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: Yep, re-ordered for clarity.

File: Sources/MongoSwift/SDAM.swift (hunk 11 above)
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: yep, added.

File: Sources/MongoSwift/SDAM.swift [hunk 13]
```diff
@@ -231,29 +254,137 @@ public struct TopologyDescription {
...
+            }
+
+            guard let clientToServer = delta(server), let clientToPrimary = delta(primary) else {
+                return nil
+            }
+
+            return clientToServer - clientToPrimary + Double(TopologyDescription.heartbeatFrequencyMS)
+        } else {
+            let possibleSMax = self.servers.filter { $0.type == .rsSecondary }.max {
+                $0.lastWriteDate!.timeIntervalSinceReferenceDate < $1.lastWriteDate!.timeIntervalSinceReferenceDate
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: For some reason I was under the impression it did not support `<`, but I'm not sure where I got that idea. I've changed it to use `<`.

Also, I added the - operator in the extension, seems to work fine.

File: Tests/MongoSwiftTests/SDAMMonitoringTests.swift [hunk 14]
```diff
@@ -118,4 +119,115 @@ final class SDAMTests: MongoSwiftTestCase {
...
+                    isMaster: isMasterSecondary,
+                    updateTime: primaryLastWrite + 600),
+            ServerDescription(
+                    connectionId: ConnectionId(hosts[1]),
+                    type: .rsSecondary,
+                    isMaster: isMasterSecondary1,
+                    updateTime: primaryLastWrite + 10)
+        ]
+
+        let serversNoPrimary: [ServerDescription] = Array(servers[1...])
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Tests/MongoSwiftTests/SDAMMonitoringTests.swift]: fixed, and yes.

File: Tests/MongoSwiftTests/SDAMMonitoringTests.swift [hunk 15]
```diff
@@ -118,4 +119,115 @@ final class SDAMTests: MongoSwiftTestCase {
...
+        ]
+
+        let isMasterSecondary1: Document = [
+            "hosts": hosts,
+            "primary": hosts[0],
+            "tags": tags1,
+            "lastWrite": lw
+        ]
+
+        let servers: [ServerDescription] = [
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Tests/MongoSwiftTests/SDAMMonitoringTests.swift]: Dang, missed these. Fixed.

File: Sources/MongoSwift/SDAM.swift [hunk 16]
```diff
@@ -94,7 +94,7 @@ public struct ServerDescription {
     public var passives: [ConnectionId] = []
 
     /// Tags for this server.
-    public var tags: [String: String] = [:]
+    public var tags: Document?
```
[the remark this answers is not in the document]
Comment (kmahar) [reply Sources/MongoSwift/SDAM.swift]: perhaps we should open a spec ticket

File: Sources/MongoSwift/SDAM.swift (hunk 13 above)
[the remark this answers is not in the document]
Comment (kmahar) [reply Sources/MongoSwift/SDAM.swift]: 🎉

File: Sources/MongoSwift/SDAM.swift [hunk 17]
```diff
@@ -231,29 +254,132 @@ public struct TopologyDescription {
...
+
+            return clientToServer - clientToPrimary + Double(TopologyDescription.heartbeatFrequencyMS)
+        } else { // ReplicaSetNoPrimary case
+            let secondaries = self.servers.filter { $0.type == .rsSecondary }
+            guard secondaries.count > 0 else {
+                return nil
+            }
+            let sMax = secondaries.max { $0.lastWriteDate! < $1.lastWriteDate! }!
+
+            return (sMax.lastWriteDate! - server.lastWriteDate!) + Double(TopologyDescription.heartbeatFrequencyMS)
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: Oh, nice catch. I do convert it for the ReplicaSetPrimary case, but forgot to do so here. I think it makes more sense if we abide by Swift's decision to use seconds as the standard time unit instead of converting to MS everywhere anyways, so I've updated all `Double`s to be `TimeIntervals` and only converted `heartbeatFrequencyMS`. The value we're comparing this against is `maxStalenessSeconds` anyways.

This should've been caught by my testing and wasn't, so I'm going to add some of the spec tests in tomorrow to make it a little more robust.

File: Sources/MongoSwift/SDAM.swift (hunk 16 above)
[the remark this answers is not in the document]
Comment (jmikola) [reply Sources/MongoSwift/SDAM.swift]: [Configure Replica Set Tag Sets](https://docs.mongodb.com/manual/tutorial/configure-replica-set-tag-sets/) may be helpful to understand this. An individual server can have any number of tags. A tag is itself a string and is associated with a string value.

This contrasts with read preference tag sets (i.e. zero or more key/value pairs), which are used for selection. A read preference can use a sequence of tag sets and will attempt to select a matching server in priority order. An empty tag set (i.e. no key/value pairs) is typically used as the last tag set in the sequence to provide a fallback of "select anything".

I think this should remain as a string:string map. @patrickfreed: Can you link to specific examples in the SS spec where `int` was used?

[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: From the [Server Selection spec](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#tag-sets):
> For example, the read preference tag set "{ dc: 'ny', rack: 2 }" matches a secondary server with tag set "{ dc: 'ny', rack: 2, size: 'large' }".

[the remark this answers is not in the document]
Comment (jmikola) [reply Sources/MongoSwift/SDAM.swift]: Interesting. That would seem to violate the definition in the server's [replica set configuration docs](https://docs.mongodb.com/manual/reference/replica-configuration/#rsconf.members[n].tags).

Would you mind opening a SPEC ticket (component: Server Selection) to report that? You might also want to submit a PR to change these to strings and possible clarify in the spec that this is a string:string mapping. I can provide some guidance on submitting a spec PR tomorrow if you like (or feel free to grab someone in the office to help today).

[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: [SPEC-1186](https://jira.mongodb.org/browse/SPEC-1186) filed. Talking tomorrow about submitting a PR sounds good to me 👍

[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: I've updated it to use a String to String map again

File: Sources/MongoSwift/SDAM.swift [hunk 18]
```diff
@@ -231,29 +254,134 @@ public struct TopologyDescription {
...
+            suitableModes = [.rsPrimary]
+        case .secondary:
+            suitableModes = [.rsSecondary]
+        default:
+            suitableModes = [.rsSecondary, .rsPrimary]
+        }
+
+        var candidates = self.servers.filter { suitableModes.contains($0.type) }
+
+        // If read preference specifies a max staleness, first filter out too stale servers
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: It does seem like we should be factoring in the latency here. I sort of read it as you did in your second paragraph, but the [read operations](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#read-operations) section of the server selection spec spells it out more explicitly.   

I'll submit another revision that checks latency.

File: Sources/MongoSwift/SDAM.swift [hunk 19]
```diff
@@ -198,6 +218,9 @@ public enum TopologyType: String {
 /// A struct describing the state of a MongoDB deployment: its type (standalone, replica set, or sharded),
 /// which servers are up, what type of servers they are, which is primary, and so on.
 public struct TopologyDescription {
+    /// The interval between server checks, can be configured via the URI.
+    internal static var heartbeatFrequencyMS: Int = 60000
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: Wow, thank you for catching this. I was under the impression that I already added the logic for updating it from the URI, but I hadn't. Will submit a patch that updates it.

File: Sources/MongoSwift/SDAM.swift [hunk 20]
```diff
@@ -167,9 +167,29 @@ public struct ServerDescription {
...
+        return true
+    }
+
+    /// An internal initializer to create a `ServerDescription` for testing purposes. This is defined here instead of
+    /// in an extension in the test files due to restrictions introduced in SE-0189.
+    internal init(connectionId: ConnectionId, type: ServerType, isMaster: Document, updateTime: Date?) {
+        self.connectionId = connectionId
+        self.type = type
+        self.parseIsMaster(isMaster)
+        self.lastUpdateTime = updateTime // TODO: get lastUpdateTime from mongoc_server_description_t after CDRIVER-2896
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: Filed [SWIFT-288](https://jira.mongodb.org/browse/SWIFT-288)

Will also update the comments with this ticket number.

File: Sources/MongoSwift/SDAM.swift (hunk 18 above)
[the remark this answers is not in the document]
Comment (jmikola) [reply Sources/MongoSwift/SDAM.swift]: Based on https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#read-operations, it doesn't look like latency is important. `maxStalenessSeconds` actually factors into narrowing down eligible secondary servers, but `localThresholdMS` is only used to prune multiple eligible servers. Per [Selecting Servers within the Latency Window](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#selecting-servers-within-the-latency-window), the smallest latency among the eligible servers becomes the baseline and you would filter out servers whose latency is too large and finally do a random selection from those that remain. That said, I don't think it will change the return value of `hasReadableServer()` at all so there may be nothing to do for `localThresholdMS`.

[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/SDAM.swift]: Oh, I see, the window is defined relative to the available servers. Then yeah, no changes need to be made to accommodate it. That might be why I left it out in the first place but I honestly don't remember.

This has made me realize that I need to change `heartbeatFrequencyMS` from being static to being stored on the `MongoClient`, though, since each client can set its own via the URI. Definitely do not want clients rewriting other clients' settings...

**Comment by patrickfreed** [comment]
After offline conversation, it was decided that this feature isn't worth duplicating and maintaining server selection code at the Swift level. If this functionality becomes a higher priority later on, this implementation can be a reference.
